Sources/MapboxMaps/Style/Style+Localization.swift

```diff
@@ -31,41 +31,61 @@ extension Style {
         }
     }
 
-    /// Filters through source to determine supported locale styles.
-    /// This is needed for v7 support
-    internal func getLocaleValue(locale: Locale) -> String? {
-        // Docs for language, region, and script codes  https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html
-        let supportedLocaleIdentifiers = ["ar", "de", "en", "es", "fr", "it", "ja", "ko", "pt", "ru", "vi", "zh", "zh-Hans", "zh-Hant", "zh-Hant-TW"]
+    /// Returns the BCP 47 language tag supported by Mapbox Streets source v8 that is most preferred according to the given preferences.
+    /// Docs for language, region, and script codes: https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html
+    internal func preferredMapboxStreetsLocalization(among preferences: [String]) -> String? {
+        let supportedLanguageCodesv8 = ["ar", "en", "es", "fr", "de", "it", "pt", "ru", "zh-Hans", "zh-Hant", "ja", "ko", "vi"].map(Locale.init(identifier:))
+
+        let preferredLocales = preferences.map(Locale.init(identifier:))
+
+        let mostSpecificLanguage = Bundle.preferredLocalizations(from: supportedLanguageCodesv8.map(\.identifier),
+            forPreferences: preferences)
+            .first
 
-        // Do nothing if we do not support the locale
-        if !supportedLocaleIdentifiers.contains(locale.languageCode!) {
+        // `Bundle.preferredLocalizations(from:forPreferences:)` is just returning the first localization it could find.
+        if let mostSpecificLanguage = mostSpecificLanguage,
+            !preferredLocales.contains(where: { $0.languageCode == Locale(identifier: mostSpecificLanguage).languageCode }) {
             return nil
         }
+        return mostSpecificLanguage
+    }
 
-        let vectorSources = allSourceIdentifiers.filter { source in
-            return source.type == .vector
+    /// Returns the shortened language identifier string representing a supported Mapbox Streets Localization
+    /// List of supported language identifiers: https://docs.mapbox.com/data/tilesets/reference/mapbox-streets-v8/#common-fields
+    internal func getLocaleValue(locale: Locale) -> String? {
+        let preferences: [String]
+        // Check if the passed Locale is the system or a created Locale
+        if locale == Locale.autoupdatingCurrent || locale == Locale.current {
+            preferences = Locale.preferredLanguages
+        } else {
+            preferences = [locale.identifier]
         }
 
-        for sourceInfo in vectorSources where locale.identifier.starts(with: "zh") {
+        // Lists language codes supported by Mapbox Streets v7
+        let supportedLanguageCodesv7 = ["ar", "en", "es", "fr", "de", "pt", "ru", "ja", "ko", "zh", "zh-Hans"]
```

Comment:
I'm confused by having v7 list in the `getLocaleValue(locale: Locale)` method and v8 list in the `preferredMapboxStreetsLocalization`.
Can we have reorganise it, please?

Reply:
We also need some space for future streets versions.

Reply:
By the way, why there is only streets style localisation codes? What is about other core/custom styles?

Reply:
These two lists reflect different support at the tileset/source level for various languages by [Streets v7](https://docs.mapbox.com/data/tilesets/reference/legacy/mapbox-streets-v7/#name-fields) and [Streets v8](https://docs.mapbox.com/data/tilesets/reference/mapbox-streets-v8/#common-fields), which would include all Mapbox core styles. In the most recent commit I added an explanation in `getLocaleValue(locale:)` and required passing the list directly to `preferredMapboxStreetsLocalization(among:, from:)` so this is more clear. 

One additional modification we could do is allow a developer to pass their own list of supported language codes (representing the set of language codes supported by their tileset). This may be outside the scope of this PR, but I can create a ticket for it.